datasets/noaa/_images/run_csv_transform_kub_gsod_stations/csv_transform.py

```diff
@@ -0,0 +1,304 @@
+    # ftp_dir               STRING          -> The remote working directory that the FTP file resides in (directory only)
+    # ftp_filename          STRING          -> The name of the file to pull from the FTP site
+    # source_file           PATHLIB.PATH    -> The (local) path pertaining to the downloaded source file
+    # target_file           PATHLIB.PATH    -> The (local) target transformed file + filename
+    # target_gcs_bucket     STRING          -> The target GCS bucket to place the output (transformed) file
+    # target_gcs_path       STRING          -> The target GCS path ( within the GCS bucket ) to place the output (transformed) file
+
+    curr_dtm = str(datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"))
+    logging.info(f"NOAA GSOD Stations By Year process started at {curr_dtm}")
+
+    curr_dtm = str(datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"))
+    logging.info(f"starting processing {source_url} at {curr_dtm}")
+
+    if url_is_reachable(source_url):
+
+        curr_dtm = str(datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"))
+        logging.info("creating 'files' folder")
+        pathlib.Path("./files").mkdir(parents=True, exist_ok=True)
+
+        curr_dtm = str(datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"))
+        logging.info(f"Downloading FTP file {source_url} from {ftp_host}")
+        download_file_ftp(ftp_host, ftp_dir, ftp_filename, source_file, source_url)
+
+        curr_dtm = str(datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"))
+        logging.info(f"Removing unnecessary header in {source_file} at {curr_dtm}")
+        os.system(f"tail -n +21 {source_file}.bak > {source_file}.1")
+        os.system(f"sed '2d' {source_file}.1 > {source_file}")
+        os.unlink(str(source_file) + ".bak")
+        os.unlink(str(source_file) + ".1")
+
+        curr_dtm = str(datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"))
+        logging.info(f"Opening source file {source_file} at {curr_dtm}")
+        colspecs = [
+            (0, 6),
+            (7, 12),
+            (13, 42),
+            (43, 45),
+            (48, 50),
+            (51, 56),
+            (57, 64),
+            (65, 74),
+            (75, 81),
+            (82, 90),
+            (91, 99),
+        ]
+        df = pd.read_fwf(str(source_file), colspecs=colspecs)
+
+        curr_dtm = str(datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"))
+        logging.info(f"Transform: Renaming Headers.. {source_file}")
+        df.columns = [
+            "usaf",
+            "wban",
+            "name",
+            "country",
+            "state",
+            "call",
+            "lat",
+            "lon",
+            "elev",
+            "begin",
+            "end",
+        ]
+
+        # remove rows with empty (usaf) data
+        df = df[df.usaf != ""]
+
+        # execute reg-ex replacements
+        curr_dtm = str(datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"))
```

Comment:
Won't need this due to built-in monitoring

Reply:
Done